### ISSUE TITLE
Random stuck in eclipse

### ISSUE THREAD

**Issue by qinsoon** [mmtk/mmtk-jikesrvm#36]
JikesRVM may randomly get stuck in eclipse (maybe in other benchmarks as well). 
* After the workpacket refactoring, it happens more frequently. 
* With 1 GC thread, it almost always happens.

**Comment by qinsoon** [comment]
This actually happens for a lot of benchmarks. It can be easily reproduced by running with 1 GC thread. 

What appears to happen is that when MMTk requires to stop all the mutators, JikesRVM gets stuck at this line `t.monitor().lockNoHandshake()` in `RVMThread.blockAllMutatorsForGC()` (https://github.com/JikesRVM/JikesRVM/blob/682344f6eb9b9c3a75b8ad54b62ceb4b133b94f5/rvm/src/org/jikesrvm/scheduler/RVMThread.java#L3985). It fails to acquire a lock for `OSROrganizerThread`, which should have released its lock and is waiting at this line (https://github.com/JikesRVM/JikesRVM/blob/682344f6eb9b9c3a75b8ad54b62ceb4b133b94f5/rvm/src/org/jikesrvm/adaptive/OSROrganizerThread.java#L43). 

I tried to log more information (with `VM.sysWrite()`). It seems when I try to log at the line before the lock acquire, the stuck problem disappeared (though sometimes at a later stage, the execution may get stuck at different places, e.g. `MemoryManager.prepareMutator()`, in which, we also try lock with `monitor().lockNoHandshake()`). I could not figure out the issue. 

Disabling recompilation (`-X:aos:enable_recompilation=false`) can be a temporary fix for our tests (at least our tests won't get stuck). 

This problem exists in rev e088173b39e18a26ee4ec5fce0c2fb3e619c4964, which is before the work packet was introduced. So the problem was not introduced by the work packet changes.

**Comment by qinsoon** [comment]
Disabling recompilation does not fix this. We reverted the change about disabling recompilation in tests, https://github.com/mmtk/mmtk-jikesrvm/pull/45. We still have this issue with our JikesRVM test.

**Comment by javadamiri** [comment]
Just an update to be saved here:

When running with `RBaseBaseSemiSpace` instead of `RFastAdaptiveSemiSpace`, I see assertion failures, e.g.:
```
Died in GC:
vm internal error at:
Thread #5
-- Stack --
(0x688b48e8 14122)   at [0x688b48e8, 0x643685be] Lorg/jikesrvm/VM; sysFail(Ljava/lang/String;)V at line 2314 at bytecode index 27 at machine code offset 0x0000006c
(0x688b4908 13951)   at [0x688b4908, 0x643686cb] Lorg/jikesrvm/VM; _assertionFailure(Ljava/lang/String;Ljava/lang/String;)V at line 654 at bytecode index 34 at machine code offset 0x00000086
(0x688b4930 13950)   at [0x688b4930, 0x64368759] Lorg/jikesrvm/VM; _assert(ZLjava/lang/String;Ljava/lang/String;)V at line 637 at bytecode index 6 at machine code offset 0x0000003b
(0x688b4954 13948)   at [0x688b4954, 0x641fb3ff] Lorg/jikesrvm/VM; _assert(Z)V at line 612 at bytecode index 3 at machine code offset 0x00000029
(0x688b4974 372)   at [0x688b4974, 0x641fb555] Lorg/jikesrvm/scheduler/Monitor; lockNoHandshake()V at line 107 at bytecode index 50 at machine code offset 0x0000009b
(0x688b4990 6715)   at [0x688b4990, 0x6421b101] Lorg/jikesrvm/scheduler/NoYieldpointsMonitor; lockNoHandshake()V at line 47 at bytecode index 4 at machine code offset 0x00000029
(0x688b49b0 14413)   at [0x688b49b0, 0x6435595a] Lorg/jikesrvm/scheduler/RVMThread; unblock(Lorg/jikesrvm/scheduler/RVMThread$BlockAdapter;)V at line 2512 at bytecode index 4 at machine code offset 0x00000035
(0x688b49d8 14541)   at [0x688b49d8, 0xf7bb730f] Lorg/jikesrvm/scheduler/RVMThread; unblockAllMutatorsForGC()V at line 4022 at bytecode index 79 at machine code offset 0x0000013e
   at [0x688b4f8c, 0x6424e289] <native frame>
(0x688b4fb0 3050)   at [0x688b4fb0, 0x64217e64] Lorg/jikesrvm/runtime/SysCall; sysStartControlCollector(Lorg/vmmagic/unboxed/Address;)V at line 189 at bytecode index 2 at machine code offset 0x00000031
(0x688b4fd4 6933)   at [0x688b4fd4, 0x64357224] Lorg/jikesrvm/mm/mminterface/CollectorThread; run()V at line 101 at bytecode index 23 at machine code offset 0x00000054
(0x688b4ff4 14520)   at [0x688b4ff4, 0xf7f74bf5] Lorg/jikesrvm/scheduler/RVMThread; startoff()V at line 2798 at bytecode index 71 at machine code offset 0x00000128
```

These assertions are not run in the fast build, so the assertion failures probably propagates to other places in the fast build.

**Comment by javadamiri** [comment]
I think this issue can be closed as a result of PR #61
